src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java

```diff
@@ -300,17 +308,42 @@ public void publish(ClusterState clusterState, final Discovery.AckListener ackLi
         return members.toArray(new LocalDiscovery[members.size()]);
     }
 
-    private void publish(LocalDiscovery[] members, ClusterState clusterState, final BlockingClusterStatePublishResponseHandler publishResponseHandler) {
+    private void publish(LocalDiscovery[] members, ClusterChangedEvent clusterChangedEvent, final BlockingClusterStatePublishResponseHandler publishResponseHandler) {
 
         try {
             // we do the marshaling intentionally, to check it works well...
-            final byte[] clusterStateBytes = Builder.toBytes(clusterState);
+            byte[] clusterStateBytes = null;
+            byte[] clusterStateDiffBytes = null;
 
+            ClusterState clusterState = clusterChangedEvent.state();
             for (final LocalDiscovery discovery : members) {
                 if (discovery.master) {
                     continue;
                 }
-                final ClusterState nodeSpecificClusterState = ClusterState.Builder.fromBytes(clusterStateBytes, discovery.localNode);
+                ClusterState newNodeSpecificClusterState = null;
+                synchronized (lock) {
```

Comment:
Sorry, I am just always irked by the `synchronized` and `synchronized(this)` construct. They basically expose something as intimate as internal concurrency handling logic to the entire world - basically if you have a pointer to the object you can mess up with its concurrency primitives. But since we are using this approach everywhere else I will change it.